src/cpuset.rs

```diff
@@ -0,0 +1,217 @@
+        CpuSet { mask: mask }
+    }
+
+    /// Create a new `CpuSet` from a given mask. For example a u64 or a u8.
+    pub fn from_mask<T>(mask: T) -> CpuSet {
+        let mut cpuset = Self::new(8 * mem::size_of::<T>());
+        unsafe { ptr::write(cpuset.mut_mask_ptr() as *mut T, mask); }
+        cpuset
+    }
+
+    /// Create a new `CpuSet` that with one CPU set as active.
+    /// Shorthand for using `new` and `set`
+    pub fn single(cpu: usize) -> CpuSet {
+        let mut cpuset = Self::new(cpu+1);
+        cpuset.set(cpu);
+        cpuset
+    }
+
+    /// Activate a given `cpu` on this `CpuSet`.
+    /// If the given `cpu` does not fit in the current `CpuSet`, it will be expanded to fit.
+    /// Equivalent of `CPU_SET`
+    pub fn set(&mut self, cpu: usize) {
+        let elem = cpu / MASK_BITS;
+        let bit = cpu % MASK_BITS;
+        while elem > self.mask.len() {
+            self.mask.push(0);
+        }
+        self.mask[elem] |= 1 << bit;
+    }
+
+    /// Clear a given `cpu` on this `CpuSet`.
+    /// If the given `cpu` does not fit within the current `CpuSet` nothing will happen.
+    /// Equivalent of `CPU_CLR`.
+    pub fn clear(&mut self, cpu: usize) {
+        let elem = cpu / MASK_BITS;
+        let bit = cpu % MASK_BITS;
+        if elem < self.mask.len() {
+            self.mask[elem] ^= 1 << bit;
+        }
+    }
+
+    /// Get if a given CPU is active in this `CpuSet`.
+    /// If `cpu` does not fit in this `CpuSet` false will be returned.
+    /// Equivalent of `CPU_ISSET`.
+    pub fn is_set(&self, cpu: usize) -> bool {
+        let elem = cpu / MASK_BITS;
+        let bit = cpu % MASK_BITS;
+        if elem > self.len() {
+            false
+        } else {
+            self.mask[elem] & (1 << bit) != 0
+        }
+    }
+
+    /// Get the number of bytes in the mask.
+    /// Produces the same results as `CPU_ALLOC_SIZE`.
+    pub fn len(&self) -> usize {
+        (MASK_BITS / 8) * self.mask.len()
+    }
+
+    /// Get the raw pointer to the bitmask
+    /// Any modification of the `CpuSet` after this call might invalidate the pointer.
+    pub fn mask_ptr(&self) -> *const c_void {
+        &self.mask[0] as *const _ as *const c_void
+    }
+
+    /// Get a mutable raw pointer to the bitmask.
+    /// Any modification of the `CpuSet` after this call might invalidate the pointer.
+    pub fn mut_mask_ptr(&mut self) -> *mut c_void {
+        (&mut self.mask[0]) as *mut _ as *mut c_void
+    }
+
+    /// Represent this `CpuSet` as an integer.
+    /// Will return an `Err` if the `CpuSet` is too large to be written to `T`
+    pub fn as_int<T: num::integer::Integer + num::traits::Zero>(&self) -> Result<T, ()> {
+        let src_size = self.len();
+        let out_size = mem::size_of::<T>();
+        if src_size > out_size {
+            Err(())
+        } else {
+            let mut mask: T = num::zero();
+            unsafe { ptr::copy(self.mask_ptr(), (&mut mask) as *mut _ as *mut c_void, src_size) }
```

Comment:
this is not safe, `T: num::integer::Integer` does not mean it is laid out like a CPU integer - this function should probably either return a concrete type or be marked `unsafe`